### PR TITLE
Change React.Component<P, S> to React.Component<P = {}, S = {}>

### PR DESCRIPTION
**Pull request by tkrotoff**
Allow to write `React.Component<Props>` and `React.Component` instead of `React.Component<Props, NoState>` and `React.Component<NoProps, NoState>`.
This feature requires TypeScript 2.3.

Read this: https://blogs.msdn.microsoft.com/typescript/2017/04/27/announcing-typescript-2-3/#default-type-arguments

This PR is big because all types that depend on React needs to be updated to TypeScript 2.3 and I've applied the simplification to all packages (a good way to test everything is ok).

4 important commits:
- [Change React.Component<P, S> to React.Component<P, S = {}>](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288/commits/3c9eba9d3f471652f0a33596deb2b4a10c74a0d5) => see ***
- [Replace React.Component<(.*), (any|void|\{\}|null)> by React.Component<$1>](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288/commits/ea9de8a518b01fa4257d367418e95a4ed2897717)
- [Change React.Component<P, ...> to React.Component<P = {}, ...>](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288/commits/b0c9072f430eb3304fb79011a36b522d54c2be9c) => props are `{}` by default, see https://codepen.io/tkrotoff/pen/WORBwX
- [Replace React.Component<(any|void|\{\}|null)> by React.Component](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288/commits/573083a5188c72df56b866b86523c9c9bf91c7ec)

There is also [a fix due to TypeScript 2.3 for react-fa](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288/commits/a42c34d876e88d2be156dcdb2deb02d38ebf9aad) and I need @flaub, @patsissons, @LKay to review it.
The other commits are cosmetic.

Once this is merged, #17021 should be easy to merge

## *** Should it be:
1. `React.Component<..., S = {}>` most common way to express no state
2. `React.Component<..., S = void>` suggested by https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8aff840533088365ab17b7be1bb1b8df3cb35e8a/types/react/test/index.ts#L133
3. `React.Component<..., S = object>` suggested by https://blogs.msdn.microsoft.com/typescript/2017/04/27/announcing-typescript-2-3/#default-type-arguments (`object` has been introduced in [TypeScript 2.2](https://blogs.msdn.microsoft.com/typescript/2017/02/22/announcing-typescript-2-2/))
4. `React.Component<..., S = null>` suggested by https://codepen.io/tkrotoff/pen/WORBwX and https://github.com/infernojs/inferno/blob/v3.5.2/packages/inferno-component/src/index.ts#L198
=> _This is how it should be done ideally!_
5. `React.Component<..., S = undefined>`

### Quick test:
```TypeScript
interface Props {
    hello: string;
}
interface State {
    foobar: string;
}
class ComponentWithPropsAndState extends React.Component<Props, State> {
}
<ComponentWithPropsAndState hello="TypeScript" />

class ComponentWithoutState extends React.Component<Props> {
}
<ComponentWithoutState hello="TypeScript" />

class ComponentWithoutPropsAndState extends React.Component {
}
<ComponentWithoutPropsAndState />

const StatelessComponentWithoutProps: React.SFC = (props) => {
    return <div />;
};
<StatelessComponentWithoutProps />;
```

### Results:

1. `{}` is OK
2. `void` fails
```
JSX element type 'ComponentWithPropsAndState' is not a constructor function for JSX elements.
  Types of property 'setState' are incompatible.
    Type '{ <K extends "foobar">(f: (prevState: State, props: Props) => Pick<State, K>, callback?: (() => a...' is not assignable to type '{ <K extends never>(f: (prevState: void, props: any) => Pick<void, K>, callback?: (() => any) | u...'.
      Types of parameters 'f' and 'f' are incompatible.
        Type '(prevState: void, props: any) => Pick<void, any>' is not assignable to type '(prevState: State, props: Props) => Pick<State, any>'.
          Types of parameters 'prevState' and 'prevState' are incompatible.
            Type 'State' is not assignable to type 'void'.
```
3. `object` is OK
4. `null` fails:
```
JSX element type 'ComponentWithPropsAndState' is not a constructor function for JSX elements.
  Types of property 'setState' are incompatible.
    Type '{ <K extends "foobar">(f: (prevState: State, props: Props) => Pick<State, K>, callback?: (() => a...' is not assignable to type '{ <K extends never>(f: (prevState: null, props: any) => Pick<null, K>, callback?: (() => any) | u...'.
      Types of parameters 'f' and 'f' are incompatible.
        Type '(prevState: null, props: any) => Pick<null, any>' is not assignable to type '(prevState: State, props: Props) => Pick<State, any>'.
          Types of parameters 'prevState' and 'prevState' are incompatible.
            Type 'State' is not assignable to type 'null'.
```
5. `undefined` fails:
```
JSX element type 'ComponentWithPropsAndState' is not a constructor function for JSX elements.
  Types of property 'setState' are incompatible.
    Type '{ <K extends "foobar">(f: (prevState: State, props: Props) => Pick<State, K>, callback?: (() => a...' is not assignable to type '{ <K extends never>(f: (prevState: undefined, props: any) => Pick<undefined, K>, callback?: (() =...'.
      Types of parameters 'f' and 'f' are incompatible.
        Type '(prevState: undefined, props: any) => Pick<undefined, any>' is not assignable to type '(prevState: State, props: Props) => Pick<State, any>'.
          Types of parameters 'prevState' and 'prevState' are incompatible.
            Type 'State' is not assignable to type 'undefined'.
```

### CODE REVIEW

**Comment by huhuanming** [comment]
👍

**Comment by honzabrecka** [comment]
👍

**Comment by MartynasZilinskas** [comment]
👍

**Comment by DaIgeb** [comment]
👍

**Review by alitaheri — APPROVED** [review]

**Comment by hodavidhara** [comment]
👍

**Review by jwbay — APPROVED** [review]
👍

**Comment by Kovensky** [comment]
The problem with this change is that, unless explicitly set, the default state for a component is `undefined`. While `setState` calls happen to work even without an existing state, they work for the same reason an object spread of `undefined` works; but it is _not_ possible to read from the state without initializing it first (giving it a type).

Is it perhaps a compiler bug that giving a default value to a template type causes giving a concrete value incompatible with the default to fail type checking?

**Comment by tkrotoff** [comment]
> the default state for a component is `undefined`
Not exactly https://codepen.io/tkrotoff/pen/WORBwX, https://github.com/infernojs/inferno/blob/v3.5.2/packages/inferno-component/src/index.ts#L198

**Comment by stkb** [comment]
👍

**Comment by seryl** [comment]
👍

**Comment by tkrotoff** [comment]
The 2 new commits from above are about making props `{}` by default (change `React.Component<P, ...>` to `React.Component<P = {}, ...>`).
See https://codepen.io/tkrotoff/pen/WORBwX

**Comment by stkb** [comment]
👍

**Comment by johnnyreilly** [comment]
Thanks!

File: types/react/index.d.ts
Comment (vsiao) [inline types/react/index.d.ts]: Does `null` work? I don't think React initializes an empty state object for you.

**Comment by vsiao** [comment]
@tkrotoff can you directly mention the react maintainers next time? I don't think the right set are notified when you change 3000 files.

File: types/react/index.d.ts
Comment (vsiao) [inline types/react/index.d.ts]: nvm, I see your comments in the description about this.

Comment (vsiao) [inline types/react/index.d.ts]: Maybe it would work if you changed `JSX.ElementClass` to `extend React.Component<any, any>`.

**Comment by wolfgangmeyers** [comment]
@tkrotoff I'm seeing many errors like this after adding `@types/react-bootstrap` to my project:
```
ERROR in [at-loader] ./node_modules/@types/react-bootstrap/index.d.ts:1076:22
    TS2314: Generic type 'ClassicComponent<P, S>' requires 2 type argument(s).

ERROR in [at-loader] ./node_modules/@types/react-bootstrap/index.d.ts:1083:26
    TS2314: Generic type 'ClassicComponent<P, S>' requires 2 type argument(s).

ERROR in [at-loader] ./node_modules/@types/react-bootstrap/index.d.ts:1098:18
    TS2314: Generic type 'ClassicComponent<P, S>' requires 2 type argument(s).
```

I have typescript 2.3.4 and webpack 3.0.0. Any ideas on how I can get this to compile? I was about to create a new issue and then noticed this PR.

**Comment by tkrotoff** [comment]
@wolfgangmeyers 
It looks like you have the latest react-bootstrap/index.d.ts:
`React.ClassicComponent<MediaListProps> // Line 1076` => 1 argument

And an old react/index.d.ts (most probably a copy-pasted react/index.d.ts somewhere in your code) that still requires 2 arguments (because there is no default values for P and S) hence the error `requires 2 type argument(s)`

Check that your react/index.d.ts contains `interface ClassicComponent<P = {}, S = {}>` at line 214

**Comment by wolfgangmeyers** [comment]
Got it. I ended up removing node_modules and re-installed all dependencies, problem vanished. Thanks!

**Comment by jacobrask** [comment]
What problem did this solve, specifically? Wouldn't it have been enough to update only React, and not the depending projects? If it did not solve any actual problems other than cosmetic changes in the type definitions themselves I suggest reverting everything expect the React change, seeing the number of issues people are experiencing with incompatible type definitions.

**Comment by tkrotoff** [comment]
@jacobrask 

> What problem did this solve, specifically?

Nothing, just syntaxic sugar

> Wouldn't it have been enough to update only React, and not the depending projects?

No, all definitions depending on React need `// TypeScript Version: 2.3` header

> I suggest reverting everything

- One day or another you need to update the definitions to later versions of TypeScript. It could have been done in this PR #17021 (and #17021 fixes a bug, not just syntaxic sugar) 
- You can use the previous versions of the definitions if you don't want to update

> seeing the number of issues people are experiencing

Could you list them? I've already replied to the ones I've found.

**Comment by NickyYo** [comment]
I can't believe this pull request was approved without any backward support... breaking changes on everything.. Everything is fucked, Jenga has finally fallen...

**Comment by Kovensky** [comment]
The point of the default argument is specifically for backward compatibility. The only issues I've seen reported is about people installing newer versions of type definitions that were updated to take advantage of this change _without_ also updating `@types/react`.

The other issue is that it broke code that was using something that doesn't `extends object` as an argument, despite there being no `extends object` constraint; but in that case, you just have to stop passing the second argument.

**Comment by NickyYo** [comment]
@Kovensky I get that but still, you should push code to production knowing that it will break everything on the internet using it that does not have the latest React defs..

**Comment by jacobrask** [comment]
I am in the same situation as @NickyYo. A lot of random stuff broke with the update. I don't even know which state I was in previously to roll back, since it was not a major version bump. I have already spent several hours for no benefit what so ever to my project.

Incompatibilities between type definition versions is the reason I abandoned TypeScript 2 years ago and only now gave it another try. Please at least take the problems this change caused seriously even if it's hard to fix after the fact.

**Comment by huhuanming** [comment]
Delete the node_moudles folder and  reinstall it.
try it?

**Comment by patsissons** [comment]
just a note to those who have run into trouble in production, `@types` do **NOT** follow semver, so you should be locking type definition versions in production (either manually or via shrinkwrap).

**Comment by tkrotoff** [comment]
@dbshell update your react/index.d.ts (delete node_modules and reinstall it)

**Comment by archy-bold** [comment]
This update has caused me so many issues and updating @types/react, removing `node_modules` and re-installing doesn't help. I also can't roll back as bug fixes have been added since this PR.

I'm using an old version of webpack, an older version of TypeScript and I feel like I'm going to have to update a large number of my dependencies to get this working...

**Comment by johnnyreilly** [comment]
> I'm using an old version of webpack, an older version of TypeScript

If you're using an older version of TypeScript you may not be able to use this change as the update requires at least TypeScript 2.3 I believe.

As has been mentioned elsewhere in this thread, you want to pin your versions when it comes to type definitions. They don't obey semver - more details on why here: 

https://blog.johnnyreilly.com/2017/02/typescript-types-and-repeatable-builds.html

**Comment by archy-bold** [comment]
@johnnyreilly I use npm 5 and it does that out of the box. My problem is that a bug fix for redux-form came after these changes, meaning I'm stuck between a rock and a hard place. I'll probably have to override the definitions manually.

**Comment by NickyYo** [comment]
This is why i'm sooooo tempted to not use Typescript anymore... i see the benefits, i love it BUT! all this hassle from an accepted pull request.... it's not worth it... There has to be a better of handling definitions...

**Comment by johnnyreilly** [comment]
@archy-bold it's not ideal. Versioning has been tricky right from the off when it comes to type definitions. It's significantly better than it was but whilst there's a way to support multiple library versions of type definitions, there isn't really the equivalent support for older versions of TypeScript. Essentially, if you want the latest type definitions you need to run the latest TypeScript. Not so much of a problem back in the early days. Now enterprises are using TypeScript and are more hesitant when it comes to "embracing the new" we're starting to realize the pain this causes. (Disclaimer: I've been bitten by this myself)

**Comment by jacobrask** [comment]
Disregarding the versioning issue, the problem would not be nearly as severe if it was introduced gradually, or after a period when people had a chance to upgrade TypeScript.

This change apparently created real issues for many people without substantial benefit, and it is unclear how to prevent something similar happening again.

**Comment by NickyYo** [comment]
@jacobrask I think the problem is egocentric developers maintaining this repo... they will be the downfall for it in the future.

**Comment by jacobrask** [comment]
I am sure everyone has good intentions, and it's not always obvious what consequence a change will have. It's happened to me countless times, and it is what code review from experienced peers is there to prevent. 

Perhaps the current maintainers will be more careful in the future, but how do we know, and will the same be true for other libraries I use? 

I consider the dependencies I add carefully by evaluating a project's health through Github issues etc. I am unsure how to go about also considering the type definition quality in this evaluation process.

**Comment by johnnyreilly** [comment]
@jacobrask - given the non-semantic changes nature of type definitions, have you considered pinning your type definition versions as suggested? With that approach nothing should break unless you actively take a new version. 

For what it's worth, we don't merge until all dependent type definitions are compiling together. That was the case on this occasion. If something wasn't covered by our tests which has burned you then please do feel free to add to our test packs to make these breakages less likely. Ultimately we're an open source project and we're only as good as the contributions we receive. We welcome all help!

**Comment by archy-bold** [comment]
To be clear, this repo has helped me immensely in getting my project running with TypeScript and integrating third party plugins. Every time an issue has sprung up previously I've found I can contribute to fix the problem for myself and others. But this change has made that process impossible without upgrading a complex application's packages; a process that I can see taking a decent amount of time. There are solutions, but I just wanted to add to the debate so everyone understands that there are edge cases here. Managing a project of this immense magnitude is difficult and everyone, I'm sure, is doing it to the best of their ability.

**Comment by NickyYo** [comment]
@archy-bold true! such a big change just to add "Glitter" not my words.. one of the reasons for this commit.

**Comment by NickyYo** [comment]
@archy-bold Also nice to see a fellow Mancunian! Small world.

**Comment by johnnyreilly** [comment]
> Every time an issue has sprung up previously I've found I can contribute to fix the problem for myself and others. But this change has made that process impossible without upgrading a complex application's packages; a process that I can see taking a decent amount of time. 

@archy-bold can I confirm that it's just the usage of TypeScript 2.3 features that has caused you the issue? I just want to be clear I understand the problem you've suffered correctly.

**Comment by archy-bold** [comment]
@johnnyreilly I can't be sure to be honest. I upgraded the individual package definitions then got errors related to the React stateless component definitions. Then I upgraded the React definitions and got a lot of other errors, I'm not sure exactly what they were but some were related to other packages. I tried upgrading their definitions, which didn't seem to fix the problem. I then tried upgrading Typescript but got a host of other errors. At which point I reverted as I don't have time to look into it this weekend. I also tried deleting node_modules and re-installing, as was suggested here. I'd have to spend time going through each error to work out if it's my code that's not playing nice with Typescript 2.4 or other packages'.

Basically my issue is that I need a fix from the new definitions of a module, but that fix comes with the changes introduced in the new definitions for React. Which in turn seems to require that I upgrade Typescript.

I'm not asking that these changes be reverted, just adding my voice to the discussion as I know others will have experienced similar. I'll either upgrade everything or override the package definitions.

@NickyYo MANCHESTER!

**Comment by johnnyreilly** [comment]
Thanks. FWIW I seem to recall that the policy for adopting new TypeScript features into DefinitelyTyped definitions is 1 month post TypeScript release. Given 2.3 had been out for about 2 months then it sounds reasonable that this PR landed when it did.

@mhegazy - I just wanted to make you aware of this discussion since I think you've previously contributed to similar discussions. I'm not sure there's anything that can usefully be done to minimise the pain of new language features landing in type definitions but if you or anyone else has any suggestions then they might be welcome.

**Pull request merged**
